pypsa/optimization/abstract.py

```diff
@@ -248,3 +249,122 @@ def optimize_security_constrained(
                 m.add_constraints(lhs, sign, rhs, name=constraint + "-security")
 
     return n.optimize.solve_model(**kwargs)
+
+
+def optimize_mga(
+    n,
+    snapshots=None,
+    multi_investment_periods=False,
+    weights=None,
+    sense="min",
+    slack=0.05,
+    model_kwargs={},
+    **kwargs,
+):
+    """
+    Run modelling-to-generate-alternatives (MGA) on network to find near-
+    optimal solutions.
+
+    Parameters
+    ----------
+    n : pypsa.Network snapshots : list-like
+        Set of snapshots to consider in the optimization. The default is None.
+    multi_investment_periods : bool, default False
+        Whether to optimise as a single investment period or to optimize in
+        multiple investment periods. Then, snapshots should be a
+        ``pd.MultiIndex``.
+    weights : dict-like
+        Weights for alternate objective function. The default is None, which
+        minimizes generation capacity. The weights dictionary should be keyed
+        with the component and variable (see ``pypsa/variables.csv``), followed
+        by a float, dict, pd.Series or pd.DataFrame for the coefficients of the
+        objective function. Examples:
+
+        >>> {"Generator": {"p_nom": 1}}
+        >>> {"Generator": {"p_nom": pd.Series(1, index=n.generators.index)}}
+        >>> {"Generator": {"p_nom": {"gas": 1, "coal": 2}}}
+        >>> {"Generator": {"p": pd.Series(1, index=n.generators.index)}
+        >>> {"Generator": {"p": pd.DataFrame(1, columns=n.generators.index, index=n.snapshots)}
+
+        Weights for non-extendable components are ignored. The dictionary does
+        not need to provide weights for all extendable components.
+    sense : str|int
+        Optimization sense of alternate objective function. Defaults to 'min'.
+        Can also be 'max'.
+    slack : float
+        Cost slack for budget constraint. Defaults to 0.05.
+    model_kwargs: dict
+        Keyword arguments used by `linopy.Model`, such as `solver_dir` or
+        `chunk`.
+    **kwargs:
+        Keyword argument used by `linopy.Model.solve`, such as `solver_name`,
+
+    Returns
+    -------
+    None
+    """
+    if snapshots is None:
+        snapshots = n.snapshots
+
+    if weights is None:
+        weights = dict(Generator=dict(p_nom=pd.Series(1, index=n.generators.index)))
+
+    # check that network has been solved
+    assert hasattr(
+        n, "objective"
+    ), "Network needs to be solved with `n.optimize()` before running MGA."
+
+    # create basic model
+    m = n.optimize.create_model(
+        snapshots=snapshots,
+        multi_investment_periods=multi_investment_periods,
+        **model_kwargs,
+    )
+
+    # build budget constraint
+    optimal_cost = (n.statistics.capex() + n.statistics.opex()).sum()
+    fixed_cost = n.statistics.capex_fixed().sum()
+    m.add_constraints(
+        m.objective + fixed_cost <= (1 + slack) * optimal_cost, name="budget"
```

Comment:
This is different to your MGA runs where fixed costs where not considered isn't it?

Reply:
Looks really nice!

Reply:
No, it's the same. I always considered the cost of previously installed capacity.